### PR TITLE
make_flux_sparsity_pattern() revision

### PR DESCRIPTION
**Pull request by vyushut**
face_has_flux_coupling may be a predicate made of multiple, complicated checks; It's better to move it after the main logic the face loop; Moreover, face_has_flux_coupling may be impossible to run if, for example, the neighbor is not an active element;

### CODE REVIEW

**Review by tjhei — APPROVED** [review]
Yes, this makes a lot of sense to me: Why check the users' predicate if we do not consider this face anyways?

Can you please add a simple test that calls this function with a predicate function that prints cell indices of the cell and its neighbor to ``deallog``?

(I am hoping somebody else can review the series of PRs Vladimir will open as we were working together)

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+// @sect3{Include files}
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:17]: Replace this comment with a short description of what this test does.

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <fstream>
+
+#include "../tests.h"
+
+
+
+// @sect3{The main() function}
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:48]: remove these comments

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  deallog << "dealii::Sparsity" << std::endl;
+  {
+    // Generate Q1 dofs for the mesh grid
+    DoFHandler<dim> dof_handler(triangulation);
+    const FE_Q<dim> finite_element(1);
+    dof_handler.distribute_dofs(finite_element);
+
+    // Compute the sparsity pattern specifying a face filter
+    DynamicSparsityPattern dsp(dof_handler.n_dofs(), dof_handler.n_dofs());
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:122]: do you also want to print the pattern?

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+main()
+{
+  initlog();
+  const int dim = 2;
+
+  // create a square with two elements
+  Triangulation<dim> triangulation;
+  make_two_elements(triangulation);
+
+  deallog << "dealii::Sparsity" << std::endl;
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:114]: can you remove this line?

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+// test that a face is visited once
+using namespace dealii;
+
+template <int dim>
+bool
+is_face_on_OY(const typename Triangulation<dim>::active_cell_iterator &cell,
+              const unsigned int face_index)
+{
+  bool on_OY = (std::abs(cell->face(face_index)->center()(0)) < 0.01);
+  deallog.get_file_stream()
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:58]:
```suggestion
  deallog
```

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    << "This sentence should appear once when the corresponding face is visited only once"
+    << std::endl;
+  deallog.get_file_stream() << cell->index() << std::endl;
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:61]:
```suggestion
  deallog << "cell index:" << cell->index() << std::endl;
```

File: tests/sparsity/flux_sparsity_pattern_visiting_once.cc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+// @sect3{Include files}
+
+#include <deal.II/base/convergence_table.h>
```
Comment (tjhei) [inline tests/sparsity/flux_sparsity_pattern_visiting_once.cc:19]: can you go through and remove some of the unnecessary headers?

**Review by masterleinad — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Review by bangerth — APPROVED** [review]
Yes, nice work!

**Pull request merged**